File: features/discovery/discovery_common/src/main/java/com/tokopedia/discovery/common/data/DynamicFilterModel.java
package com.tokopedia.discovery.common.data;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

public final class DynamicFilterModel implements Parcelable {

    @SerializedName("process_time")
    @Expose
    String processTime;
    @SerializedName("data")
    @Expose
    DataValue data = new DataValue();
    @SerializedName("status")
    @Expose
    String status;

    public DynamicFilterModel() {
    }

    public String getProcessTime() {
        return processTime;
    }

    public void setProcessTime(String processTime) {
        this.processTime = processTime;
    }

    public DataValue getData() {
        return data;
    }

    public void setData(DataValue data) {
        this.data = data;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.processTime);
        dest.writeParcelable(this.data, flags);
        dest.writeString(this.status);
    }

    protected DynamicFilterModel(Parcel in) {
        this.processTime = in.readString();
        this.data = in.readParcelable(DataValue.class.getClassLoader());
        this.status = in.readString();
    }

    public static final Creator<DynamicFilterModel> CREATOR = new Creator<DynamicFilterModel>() {
        @Override
        public DynamicFilterModel createFromParcel(Parcel source) {
            return new DynamicFilterModel(source);
        }

        @Override
        public DynamicFilterModel[] newArray(int size) {
            return new DynamicFilterModel[size];
        }
    };
}

File: features/discovery/filter/src/main/java/com/tokopedia/filter/common/data/DynamicFilterModel.kt
package com.tokopedia.filter.common.data

import android.os.Parcelable
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import com.tokopedia.discovery.common.constants.SearchApiConst
import kotlinx.android.parcel.Parcelize

@Parcelize
class DynamicFilterModel(@SerializedName("data")
                         @Expose
                         var data: DataValue = DataValue(),
                         @SerializedName("status")
                         @Expose
                         var status: String = "",
                         var defaultSortValue: String = SearchApiConst.DEFAULT_VALUE_OF_PARAMETER_SORT
) : Parcelable {

    fun isEmpty(): Boolean {
        return data.filter.map { it.options }.flatten().isEmpty()
    }

    fun hasSort() = data.sort.isNotEmpty()

    fun getSortKey() = if (hasSort()) data.sort[0].key else ""

    fun getAppliedSort(mapParameter: Map<String, Any>): Sort? {
        val sortValue = mapParameter[getSortKey()]?.toString() ?: ""

        return if (defaultSortValue == sortValue) null
        else findAppliedSort(sortValue)
    }

    private fun findAppliedSort(sortValue: String): Sort? =
        data.sort.find {
            it.key == getSortKey()
                && it.value == sortValue
        }
}
